### ISSUE TITLE
[Bug] different number of reads in paired input files; repair.sh does not seem to be working in preprocess.py

### ISSUE THREAD

**Issue by yangchen2** [jolespin/veba#27]
**Describe the bug:**
Preprocessing module is failing due to different number of reads in paired input files either at step 2 (Bowtie2) or step 3 (Bbduk). I believe this is meant to be fixed automatically as part of repair.sh in the preprocess module, but does not seem to be working. To get around the error, I ran repair.sh first and then VEBA’s preprocess.py module. 

**Versions**

```
-bash-4.2$ source activate VEBA-preprocess_env 
(VEBA-preprocess_env) -bash-4.2$ preprocess.py -v
preprocess.py v2023.2.28

```

**Command used to produce error:**

This occurs when running the preprocessing module.

**Please provide the following files:**
Two types of errors are occuring:

1. Failing at step 2 (Bowtie2):
```
cat ../veba_output/preprocess/G_1_S97_L002/log/2__bowtie2.e
Error, fewer reads in file specified with -2 than in file specified with -1
terminate called after throwing an instance of 'int'
(ERR): bowtie2-align died with signal 6 (ABRT) (core dumped)
```
2. Failing at step 3 (Bbduk):
```
cat ../veba_output/preprocess/G_2_S98_L002/log/3__bbduk.e
java -ea -Xmx85810m -Xms85810m -cp /home/yac027/mambaforge3/envs/VEBA-preprocess_env/opt/bbmap-39.01-0/current/ jgi.BBDuk zl=1 overwrite=t threads=4 in1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/2__bowtie2/cleaned_1.fastq.gz in2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/2__bowtie2/cleaned_2.fastq.gz ref=/panfs/yac027/VEBA_analyses/veba_database/Contamination/kmers/ribokmers.fa.gz k=31 minlen=75 out1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/non-kmer_hits_1.fastq.gz out2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/non-kmer_hits_2.fastq.gz outm1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/kmer_hits_1.fastq.gz outm2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/kmer_hits_2.fastq.gz
Executing jgi.BBDuk [zl=1, overwrite=t, threads=4, in1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/2__bowtie2/cleaned_1.fastq.gz, in2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/2__bowtie2/cleaned_2.fastq.gz, ref=/panfs/yac027/VEBA_analyses/veba_database/Contamination/kmers/ribokmers.fa.gz, k=31, minlen=75, out1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/non-kmer_hits_1.fastq.gz, out2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/non-kmer_hits_2.fastq.gz, outm1=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/kmer_hits_1.fastq.gz, outm2=/panfs/yac027/VEBA_analyses/Tatsuya_shotgun/veba_output/preprocess/G_2_S98_L002/intermediate/3__bbduk/kmer_hits_2.fastq.gz]
Version 39.01

Set threads to 4
0.046 seconds.
Initial:
Memory: max=89992m, total=89992m, free=89925m, used=67m

Added 12672418 kmers; time: 	5.325 seconds.
Memory: max=89992m, total=89992m, free=88785m, used=1207m

Input is being processed as paired
Started output streams:	0.403 seconds.
[E::bgzf_read_block] [E::bgzf_read_block] Invalid BGZF header at offset 210118798Invalid BGZF header at offset 202178030

[E::bgzf_read] [E::bgzf_read] Read block operation failed with error 6 after 23296 of 65536 bytesRead block operation failed with error 6 after 38912 of 65536 bytes

Error 7 in block starting at offset 209941143(C837297)
Error 7 in block starting at offset 202013184(C0A7A00)
java.lang.AssertionError: 
There appear to be different numbers of reads in the paired input files.
The pairing may have been corrupted by an upstream process. It may be fixable by running repair.sh.
	at stream.ConcurrentGenericReadInputStream.pair(ConcurrentGenericReadInputStream.java:499)
	at stream.ConcurrentGenericReadInputStream.readLists(ConcurrentGenericReadInputStream.java:364)
	at stream.ConcurrentGenericReadInputStream.run0(ConcurrentGenericReadInputStream.java:208)
	at stream.ConcurrentGenericReadInputStream.run(ConcurrentGenericReadInputStream.java:184)
	at [java.base/java.lang.Thread.run(Thread.java:833)](url)
```

Also, please provide the input files used so I can reproduce the error and further diagnose (I'll be able to help you better).  In the meantime, I am currently working on more useful error logs so patience is appreciated.

**Comment by jolespin** [comment]
Can you copy/paste the context here?
```
conda list --name VEBA-preprocess_env
```

**Comment by yangchen2** [comment]
```
conda list --name VEBA-preprocess_env
# packages in environment at /home/yac027/mambaforge3/envs/VEBA-preprocess_env:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
alsa-lib                  1.2.8                h166bdaf_0    conda-forge
argparse-manpage-birdtools 1.7.0              pyhd8ed1ab_0    conda-forge
aria2                     1.36.0               h8b6cd97_3    conda-forge
arrow-cpp                 10.0.1           h3e2b116_1_cpu    conda-forge
aws-c-auth                0.6.21               h3cb7b9d_0    conda-forge
aws-c-cal                 0.5.20               hd3b2fe5_3    conda-forge
aws-c-common              0.8.5                h166bdaf_0    conda-forge
aws-c-compression         0.2.16               hf5f93bc_0    conda-forge
aws-c-event-stream        0.2.15              h2c1f3d0_11    conda-forge
aws-c-http                0.6.27               hb11a807_3    conda-forge
aws-c-io                  0.13.11              hf1b0a34_1    conda-forge
aws-c-mqtt                0.7.13               h93e60df_9    conda-forge
aws-c-s3                  0.1.51              h1222a00_14    conda-forge
aws-c-sdkutils            0.1.7                hf5f93bc_0    conda-forge
aws-checksums             0.1.13               hf5f93bc_5    conda-forge
aws-crt-cpp               0.18.16              hb1454fd_1    conda-forge
aws-sdk-cpp               1.9.379              hdc6349a_5    conda-forge
awscli                    1.27.23          py39hf3d152e_0    conda-forge
bbmap                     39.01                h5c4e2a8_0    bioconda
bird_tool_utils_python    0.4.1              pyhdfd78af_0    bioconda
botocore                  1.29.23            pyhd8ed1ab_0    conda-forge
bowtie2                   2.5.0            py39h3321a2d_0    bioconda
brotlipy                  0.7.0           py39hb9d737c_1005    conda-forge
bz2file                   0.98                       py_0    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2022.9.24            ha878542_0    conda-forge
cairo                     1.16.0            ha61ee94_1014    conda-forge
certifi                   2022.9.24          pyhd8ed1ab_0    conda-forge
cffi                      1.15.1           py39he91dace_2    conda-forge
charset-normalizer        2.1.1              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
cryptography              38.0.4           py39hd97740a_0    conda-forge
curl                      7.86.0               h7bff187_1    conda-forge
docutils                  0.16             py39hf3d152e_3    conda-forge
expat                     2.5.0                h27087fc_0    conda-forge
extern                    0.4.1                      py_0    conda-forge
fastp                     0.23.2               h5f740d0_3    bioconda
fastq_preprocessor        2022.4.2                   py_0    jolespin
font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.14.1               hc2a2eb6_0    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
freetype                  2.12.1               hca18f0e_1    conda-forge
genopype                  2021.8.18                  py_0    jolespin
gettext                   0.21.1               h27087fc_0    conda-forge
gflags                    2.2.2             he1b5a44_1004    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
glog                      0.6.0                h6f12383_0    conda-forge
graphite2                 1.3.13            h58526e2_1001    conda-forge
harfbuzz                  5.3.0                h418a68e_0    conda-forge
hdf5                      1.12.1          nompi_h2386368_104    conda-forge
htslib                    1.16                 h6bc39ce_0    bioconda
icu                       70.1                 h27087fc_0    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
isa-l                     2.30.0               ha770c72_4    conda-forge
jmespath                  1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h166bdaf_2    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
kingfisher                0.1.0              pyh7cba7a3_1    bioconda
krb5                      1.19.3               h3790be6_0    conda-forge
lcms2                     2.14                 h6ed2654_0    conda-forge
ld_impl_linux-64          2.39                 hcc3a1bd_1    conda-forge
lerc                      4.0.0                h27087fc_0    conda-forge
libabseil                 20220623.0      cxx17_h48a1fff_5    conda-forge
libarrow                  10.0.1           hcf5dfb8_1_cpu    conda-forge
libblas                   3.9.0           16_linux64_openblas    conda-forge
libbrotlicommon           1.0.9                h166bdaf_8    conda-forge
libbrotlidec              1.0.9                h166bdaf_8    conda-forge
libbrotlienc              1.0.9                h166bdaf_8    conda-forge
libcblas                  3.9.0           16_linux64_openblas    conda-forge
libcrc32c                 1.1.2                h9c3ff4c_0    conda-forge
libcups                   2.3.3                h3e49a29_2    conda-forge
libcurl                   7.86.0               h7bff187_1    conda-forge
libdeflate                1.13                 h166bdaf_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libevent                  2.1.10               h9b69904_4    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 12.2.0              h65d4601_19    conda-forge
libgfortran-ng            12.2.0              h69a702a_19    conda-forge
libgfortran5              12.2.0              h337968e_19    conda-forge
libglib                   2.74.1               h606061b_1    conda-forge
libgomp                   12.2.0              h65d4601_19    conda-forge
libgoogle-cloud           2.5.0                hcb5eced_0    conda-forge
libgrpc                   1.49.1               h05bd8bd_1    conda-forge
libhwloc                  2.8.0                h32351e8_1    conda-forge
libiconv                  1.17                 h166bdaf_0    conda-forge
liblapack                 3.9.0           16_linux64_openblas    conda-forge
libnghttp2                1.47.0               hdcd2b5c_1    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.21          pthreads_h78a6416_3    conda-forge
libpng                    1.6.39               h753d276_0    conda-forge
libprotobuf               3.21.10              h6239696_0    conda-forge
libsqlite                 3.40.0               h753d276_0    conda-forge
libssh2                   1.10.0               haa6b8db_3    conda-forge
libstdcxx-ng              12.2.0              h46fd767_19    conda-forge
libthrift                 0.16.0               h491838f_2    conda-forge
libtiff                   4.4.0                h0e0dad5_3    conda-forge
libutf8proc               2.8.0                h166bdaf_0    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libwebp-base              1.2.4                h166bdaf_0    conda-forge
libxcb                    1.13              h7f98852_1004    conda-forge
libxml2                   2.9.14               h22db469_4    conda-forge
libzlib                   1.2.13               h166bdaf_4    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
ncbi-ngs-sdk              2.9.0                         0    bioconda
ncurses                   6.3                  h27087fc_1    conda-forge
numpy                     1.23.5           py39h3d75532_0    conda-forge
openjdk                   17.0.3               hafdced1_4    conda-forge
openssl                   1.1.1s               h0b41bf4_1    conda-forge
orc                       1.8.0                h09e0d61_0    conda-forge
ossuuid                   1.6.2             hf484d3e_1000    conda-forge
pandas                    1.5.2            py39h4661b88_0    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
pathlib2                  2.3.7.post1      py39hf3d152e_2    conda-forge
pbzip2                    1.1.13                        0    conda-forge
pcre2                     10.40                hc3806b6_0    conda-forge
perl                      5.32.1          2_h7f98852_perl5    conda-forge
perl-alien-build          2.48            pl5321hec16e2b_0    bioconda
perl-alien-libxml2        0.17            pl5321hec16e2b_0    bioconda
perl-business-isbn        3.007           pl5321hdfd78af_0    bioconda
perl-business-isbn-data   20210112.006    pl5321hdfd78af_0    bioconda
perl-capture-tiny         0.48            pl5321hdfd78af_2    bioconda
perl-carp                 1.38            pl5321hdfd78af_4    bioconda
perl-constant             1.33            pl5321hdfd78af_2    bioconda
perl-data-dumper          2.183           pl5321hec16e2b_1    bioconda
perl-encode               3.19            pl5321hec16e2b_1    bioconda
perl-exporter             5.72            pl5321hdfd78af_2    bioconda
perl-extutils-makemaker   7.64            pl5321hd8ed1ab_0    conda-forge
perl-ffi-checklib         0.28            pl5321hdfd78af_0    bioconda
perl-file-chdir           0.1010          pl5321hdfd78af_3    bioconda
perl-file-path            2.18            pl5321hd8ed1ab_0    conda-forge
perl-file-temp            0.2304          pl5321hd8ed1ab_0    conda-forge
perl-file-which           1.24            pl5321hd8ed1ab_0    conda-forge
perl-importer             0.026           pl5321hdfd78af_0    bioconda
perl-mime-base64          3.16            pl5321hec16e2b_2    bioconda
perl-parent               0.236           pl5321hdfd78af_2    bioconda
perl-path-tiny            0.122           pl5321hdfd78af_0    bioconda
perl-pathtools            3.75            pl5321hec16e2b_3    bioconda
perl-scope-guard          0.21            pl5321hdfd78af_3    bioconda
perl-sub-info             0.002           pl5321hdfd78af_1    bioconda
perl-term-table           0.016           pl5321hdfd78af_0    bioconda
perl-test2-suite          0.000145        pl5321hdfd78af_0    bioconda
perl-uri                  5.12            pl5321hdfd78af_0    bioconda
perl-xml-libxml           2.0207          pl5321h661654b_0    bioconda
perl-xml-namespacesupport 1.12            pl5321hdfd78af_1    bioconda
perl-xml-sax              1.02            pl5321hdfd78af_1    bioconda
perl-xml-sax-base         1.09            pl5321hdfd78af_1    bioconda
pigz                      2.6                  h27826a3_0    conda-forge
pip                       22.3.1             pyhd8ed1ab_0    conda-forge
pixman                    0.40.0               h36c2ea0_0    conda-forge
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
pyarrow                   10.0.1          py39h33d4778_1_cpu    conda-forge
pyasn1                    0.4.8                      py_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 22.1.0             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
python                    3.9.15          h47a2c10_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-tzdata             2022.7             pyhd8ed1ab_0    conda-forge
python_abi                3.9                      3_cp39    conda-forge
pytz                      2022.6             pyhd8ed1ab_0    conda-forge
pytz-deprecation-shim     0.1.0.post0      py39hf3d152e_3    conda-forge
pyyaml                    5.4.1            py39hb9d737c_4    conda-forge
re2                       2022.06.01           h27087fc_1    conda-forge
readline                  8.1.2                h0f457ee_0    conda-forge
requests                  2.28.1             pyhd8ed1ab_1    conda-forge
rsa                       4.7.2              pyh44b312d_0    conda-forge
s2n                       1.3.28               h8d01263_0    conda-forge
s3transfer                0.6.0              pyhd8ed1ab_0    conda-forge
samtools                  1.16.1               h6899075_1    bioconda
scandir                   1.10.0           py39hb9d737c_6    conda-forge
seqkit                    2.3.1                h9ee0642_0    bioconda
setuptools                65.5.1             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
snappy                    1.1.9                hbd366e4_2    conda-forge
soothsayer_utils          2022.2.9                   py_0    jolespin
sra-tools                 3.0.0           pl5321hd0d85c6_1    bioconda
sracat                    0.2                  h9f5acd7_1    bioconda
tbb                       2021.7.0             h924138e_1    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tqdm                      4.64.1             pyhd8ed1ab_0    conda-forge
tzdata                    2022g                h191b570_0    conda-forge
tzlocal                   4.2              py39hf3d152e_2    conda-forge
urllib3                   1.26.13            pyhd8ed1ab_0    conda-forge
wheel                     0.38.4             pyhd8ed1ab_0    conda-forge
xorg-fixesproto           5.0               h7f98852_1002    conda-forge
xorg-inputproto           2.3.2             h7f98852_1002    conda-forge
xorg-kbproto              1.0.7             h7f98852_1002    conda-forge
xorg-libice               1.0.10               h7f98852_0    conda-forge
xorg-libsm                1.2.3             hd9c2040_1000    conda-forge
xorg-libx11               1.7.2                h7f98852_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xorg-libxext              1.3.4                h7f98852_1    conda-forge
xorg-libxfixes            5.0.3             h7f98852_1004    conda-forge
xorg-libxi                1.7.10               h7f98852_0    conda-forge
xorg-libxrender           0.9.10            h7f98852_1003    conda-forge
xorg-libxtst              1.2.3             h7f98852_1002    conda-forge
xorg-recordproto          1.14.2            h7f98852_1002    conda-forge
xorg-renderproto          0.11.1            h7f98852_1002    conda-forge
xorg-xextproto            7.3.0             h7f98852_1002    conda-forge
xorg-xproto               7.0.31            h7f98852_1007    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
yaml                      0.2.5                h7f98852_2    conda-forge
zlib                      1.2.13               h166bdaf_4    conda-forge
zstd                      1.5.2                h6239696_4    conda-forge
```

**Comment by jolespin** [comment]
I need to test the functionality of `fastp` with input files that have different numbers of reads.  The ones I've tested it on have been fresh off the sequencer so I never thought of testing it. If I need to update it, it should be a simple fix with running `repair.sh` before and then piping that into stdin of `fastp`.  I could whip out that `fastq_preprocessor` update very quickly and it would simple for you to make that patch update on your install.

**Comment by jolespin** [comment]
Addressed in this commit: 
https://github.com/jolespin/fastq_preprocessor/commit/140ccf10645711e3834c963d075eb89d41eeedba

Available in `fastq_preprocessor v2023.7.24`
